### PR TITLE
persist: allow TraceBatchMeta to hold multiple keys

### PR DESCRIPTION
**Pull request by ruchirK**
This commit allows the metadata for a trace batch to hold multiple keys (or zero
keys). Future work will partition trace batch data across multiple batches.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR adds a release note for any [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

File: src/persist/src/indexed/columnar.rs
Comment (ruchirK) [inline src/persist/src/indexed/columnar.rs:571]: remove

File: src/persist/src/indexed/encoding.rs
Comment (danhhz) [inline src/persist/src/indexed/encoding.rs:180]: explicitly call out that this is allowed to be empty

File: src/persist/src/indexed/encoding.rs
Comment (danhhz) [inline src/persist/src/indexed/encoding.rs:954]: can prob also be empty vec

File: src/persist/src/indexed/background.rs
Comment (ruchirK) [reply src/persist/src/indexed/background.rs:315]: No I think we changed res.merged_keys to `vec![MERGED_KEY]` so that we could assert on res, but we need the original list of keys (before this change the original key) to get the actual batch. When I make the change that was requested, Blob complains that no key named `MERGED_KEY` exists in the blob.

**Comment by ruchirK** [comment]
I responded to all of the pr comments lmk what you think!

File: src/persist/src/indexed/background.rs
Comment (danhhz) [reply src/persist/src/indexed/background.rs]: okay, sure, ignore that specific suggestion then. to clarify my original comment, I think it's possible to avoid the [0], let's figure out a way to do that

Comment (ruchirK) [reply src/persist/src/indexed/background.rs]: done!

**Comment by ruchirK** [comment]
@danhhz made the refactors we discussed over slack

BlobTraceBatch -> BlobTraceBatchPart
ProtoTraceBatchInline -> ProtoTraceBatchPartInline
in memory and stored representations grow an index field and the code has caveats that there may be other parts out there with the same desc
changed the test to iterate through all keys and merge data into one vec before comparing them.

In the new multipart trace batch world I can think of several new invariants:

(in no particular order)
no ((key, val), time) is duplicated across parts
data is sorted across parts by index -- in other words all of the data in part i is < any of the data at part i + 1
all parts have the same desc
no parts have the same index or the same blob key
the keys are stored in TraceBatchMeta in index order
all trace parts exist for the corresponding keys in trace metadata, and none are deleted

its not obvious to me the best way to structure validations for some of these invariants as they require access to all of the trace batch parts at once. I'm kind of inclined to say for now while we don't really use multipart batches that the temporary invariants are:

TraceBatchMeta's list of keys either has a single key or no keys
BlobTraceBatchPart's index is always 0.

this way we can merge this pr fast and unblock your work thats depending on it. If merging this pr fast is not a big concern then I can type up the new validations before merging

**Review by danhhz — APPROVED** [review]
LGTM

**Comment by ruchirK** [comment]
tftr! I'm going to rebase, squash and merge on green

**Pull request merged**
